common/src/main/java/tachyon/replay/ReplayCache.java

```diff
@@ -0,0 +1,192 @@
+  private static final long DEFAULT_MAX_SIZE = 10000;
+  /** The number of milliseconds a key may remain unaccessed in the cache before it is evicted. */
+  private static final long DEFAULT_EXPIRE_MS = 2000;
+
+  private final Cache<String, V> mCache;
+
+  /**
+   * Creates a cache with default maximum size and key expiration time.
+   *
+   * @return a cache with the default maximum size and expiration time
+   */
+  public static <V> ReplayCache<V> newReplayCache() {
+    return new ReplayCache<V>();
+  }
+
+  /**
+   * Creates a cache with the given maximum size and key expiration time.
+   *
+   * @param maxSize the maximum number of elements the cache may hold
+   * @param expireMs the amount of time to hold entries before they expire
+   * @return a cache with the specified maximum size and expiration time
+   */
+  public static <V> ReplayCache<V> newReplayCache(long maxSize, long expireMs) {
+    return new ReplayCache<V>(maxSize, expireMs);
+  }
+
+  /**
+   * Creates a cache with default maximum size and key expiration time.
+   */
+  private ReplayCache() {
+    this(DEFAULT_MAX_SIZE, DEFAULT_EXPIRE_MS);
+  }
+
+  /**
+   * Creates a cache with the given maximum size and key expiration time.
+   *
+   * @param maxSize the maximum number of elements the cache may hold
+   * @param expireMs the amount of time to hold entries before they expire
+   */
+  private ReplayCache(long maxSize, long expireMs) {
+    // TODO(andrew) Make it possible for users to configure this cache via
+    // CacheBuilder<Object, Object> from(String spec)
+    mCache = CacheBuilder.newBuilder().maximumSize(maxSize)
+        .expireAfterWrite(expireMs, TimeUnit.MILLISECONDS).<String, V>build();
+  }
+
+  /**
+   * The RPC handler method to be executed in {@link #run(String, ReplayCallable)}.
+   *
+   * @param <V> the return type of {@link #call()}
+   */
+  public interface ReplayCallable<V> {
+    V call() throws TachyonException;
+  }
+
+  /**
+   * Same with {@link ReplayCallable} except that this handler method throws
+   * {@link IOException} and is to be executed in {@link #run(String,
+   * ReplayCallableThrowsIOException}.
+   *
+   * @param <V> the return type of {@link #call()}
+   */
+  public interface ReplayCallableThrowsIOException<V> {
+    V call() throws TachyonException, IOException;
+  }
+
+  /**
+   * Handles the replay logic for the given RPC handler. The provided rpcId is used to manage
+   * replays by returning their previous return value instead of re-executing the RPC handler. This
+   * is neccessary when the RPC is non-idempotent.
+   *
+   * @param rpcId the id for the RPC
+   * @param replayCallable the handler logic for the RPC
+   * @return the result of executing replayCallable, or the cached value from a previous invocation
+   * @throws TachyonTException when {@link TachyonException} is thrown by the handler call
+   */
+  public V run(String rpcId, final ReplayCallable<V> replayCallable) throws TachyonTException {
```

Comment:
in tests you use "key" instead of "id" ... I like key better, so I would rename the argument name to simply "key"